packages/tpetra/CMakeLists.txt

```diff
@@ -24,6 +24,173 @@ IF (NOT ${PROJECT_NAME}_ENABLE_CXX11 AND NOT TPL_ENABLE_CUDA)
 ENDIF ()
 
+# As of 10 Aug 2017, Tpetra requires CUDA >= 7.5.
+ASSERT_DEFINED(TPL_ENABLE_CUDA)
+IF (TPL_ENABLE_CUDA AND DEFINED CUDA_VERSION)
+  # Don't check a variable's value unless we know it is defined.
+  IF (CUDA_VERSION VERSION_LESS "7.5")
+     MESSAGE(FATAL_ERROR "If building with CUDA, Tpetra requires at least CUDA 7.5, and preferably CUDA >= 8.0.  Your CUDA_VERSION is ${CUDA_VERSION}.  For details, please refer to Trilinos issue #1278: https://github.com/trilinos/Trilinos/issues/1278")
+  ENDIF ()
+ENDIF ()
+
+# Ask the MPI implementation if it is CUDA aware.  See #1571.
+#
+# If users have already set Tpetra_ASSUME_CUDA_AWARE_MPI, we just
+# accept their setting.  This helps with cross-compilation.
+# Otherwise, we try to figure it out for them.
+#
+# Currently, we only have logic to figure it out for them if using
+# OpenMPI >= 1.7.4.  In that case, then we may ask ompi_info:
+#
+# https://www.open-mpi.org/faq/?category=runcuda
+#
+# In order to do so, we need to find the correct ompi_info executable,
+# that is, the one that goes with the version of MPI that we are
+# actually using.  This is not necessarily the version whose
+# executables are in our PATH!  OpenMPI likes to install ompi_info in
+# the same directory as its compiler wrappers, so we can just use the
+# path of MPI_CXX_COMPILER.  If that isn't defined, we can try
+# ${MPI_BIN_DIR} or "${MPI_BASE_DIR}/bin".  Otherwise, we'll make the
+# safe, but possibly less performant assumption that MPI is not CUDA
+# aware.
+
+SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT OFF)
+
+ASSERT_DEFINED (TPL_ENABLE_CUDA)
+ASSERT_DEFINED (TPL_ENABLE_MPI)
+
+MESSAGE (STATUS "Determine whether Tpetra will assume that MPI is CUDA aware:")
+
+IF (NOT TPL_ENABLE_CUDA)
+  MESSAGE (STATUS "  - TPL_ENABLE_CUDA is OFF, so we assume that MPI is not CUDA aware.")
+  SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT OFF)
+ELSEIF (NOT TPL_ENABLE_MPI)
+  # Even if CUDA is enabled but MPI is not, Teuchos::SerialComm may
+  # assume host access of input and output buffers (e.g.,
+  # Teuchos::reduceAll on a SerialComm may copy input into output on
+  # host).  Thus, it's safer to assume that "MPI" (rather,
+  # Teuchos::SerialComm, a kind of MPI stub) can't access CUDA memory.
+  MESSAGE (STATUS "  - TPL_ENABLE_CUDA is ON but TPL_ENABLE_MPI is OFF, so we assume that (nonexistent) MPI is not CUDA aware.")
+  SET (Tpetra_ASSUME_CUDA_AWARE_MPI_DEFAULT OFF)
+ELSEIF (DEFINED Tpetra_ASSUME_CUDA_AWARE_MPI)
+  # The user set Tpetra_ASSUME_CUDA_AWARE_MPI explicitly.
+  # Just use that value as the default.
+
+  IF (Tpetra_ASSUME_CUDA_AWARE_MPI)
```

Comment:
Before we set up the option, either the variable is undefined, or users set it explicitly.  I think you're right, though, that it's not necessary to set a default value if the user has already set it.  I would like some status messages regardless, though.